src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```
@@ -329,39 +382,59 @@ public static LockFileValidityWithMatchedResults IsLockFileStillValid(PackagesLo
             {
-                return true;
+                return (true,
+                           string.Format(
+                               CultureInfo.CurrentCulture,
+                               Strings.PackagesLockFile_PackageReferencesHaveChanged,
+                               nuGetFramework.GetShortFolderName(),
+                               string.Join(", ", lockFileDependencies.Select(e => e.Id + ":" + e.RequestedVersion.ToNormalizedString())),
+                               string.Join(", ", newPackageDependencies.Select(e => e.LibraryRange.Name + ":" + e.LibraryRange.VersionRange.ToNormalizedString())))
+                           );
             }
 
             foreach (var dependency in newPackageDependencies)
             {
                 var lockFileDependency = lockFileDependencies.FirstOrDefault(d => StringComparer.OrdinalIgnoreCase.Equals(d.Id, dependency.Name));
 
-                if (lockFileDependency == null || !EqualityUtility.EqualsWithNullCheck(lockFileDependency.RequestedVersion, dependency.LibraryRange.VersionRange))
+                if (lockFileDependency == null)
+                {
+                    // dependency has changed and lock file is out of sync.
+                    return (true,
+                               string.Format(
+                                   CultureInfo.CurrentCulture,
+                                   Strings.PackagesLockFile_PackageReferenceAdded,
+                                   dependency.Name,
+                                   nuGetFramework.GetShortFolderName())
+                            );
+                }
+                if (!EqualityUtility.EqualsWithNullCheck(lockFileDependency.RequestedVersion, dependency.LibraryRange.VersionRange))
                 {
                     // dependency has changed and lock file is out of sync.
-                    return true;
+                    return (true,
+                               string.Format(
```

one thing that I didn't love when I did this was the fact that this is fails fast.

---

This comment is like Andy's comment above. I will add my response in that comment to merge threads.

---

How do the log messages look after this change?
Are they still usable?